src/components/shared/AddToPortfolioButton.tsx

```diff
@@ -0,0 +1,258 @@
+type Actions =
+  | InitAction
+  | SetPortfoliosAction
+  | SetOpenAction
+  | SetLoadingAction
+  | SetErrorAction
+  | OpenSnackbarAction
+  | CloseSnackbarAction;
+
+function reducer(state: State, action: Actions): State {
+  switch (action.type) {
+    case 'init':
+      return initialState;
+    case 'setPortfolios':
+      return {
+        ...state,
+        portfolios: action.payload,
+      };
+    case 'setOpen':
+      return {
+        ...state,
+        open: action.payload,
+      };
+    case 'setLoading':
+      return {
+        ...state,
+        loading: action.payload,
+      };
+    case 'setError':
+      return {
+        ...state,
+        error: action.payload,
+      };
+    case 'openSnackbar':
+      return {
+        ...state,
+        snackbars: {
+          ...state.snackbars,
+          [action.payload]: true,
+        },
+      };
+    case 'closeSnackbar':
+      return {
+        ...state,
+        snackbars: {
+          ...state.snackbars,
+          [action.payload]: false,
+        },
+      };
+    default:
+      throw new Error();
+  }
+}
+
+// returns the add to portfolio button and all subcomponents including the dialog window
+const AddToPortfolioButton: React.FC<AddToPortfolioButtonProps> = ({
+  symbol,
+  token,
+}) => {
+  // style and translation hooks
+  const classes = useStyles();
+  const { t } = useTranslation();
+
+  const [state, dispatch] = React.useReducer<Reducer<State, Actions>>(
+    reducer,
+    initialState
+  );
+
+  // reset state on stock change
+  React.useEffect(() => dispatch({ type: 'init' }), [symbol]);
+
+  const isMounted = React.useRef(true);
+  React.useEffect(
+    () => () => {
+      isMounted.current = false;
+    },
+    []
+  );
+
+  const openDialog = async () => {
+    dispatch({ type: 'setLoading', payload: true });
+    try {
+      // const p = await API.stock(token, symbol);
+      const p: API.PortfolioStock[] = [
+        { id: 'MOCK', qty: 2.5, virtual: true, name: 'mock portfolio' },
+      ];
+      // const p: API.PortfolioStock[] = [];
+      if (isMounted.current) {
+        dispatch({ type: 'setPortfolios', payload: p });
+        if (p.length) {
+          dispatch({ type: 'setOpen', payload: true });
+        } else {
+          // openDialog is called on button click which can trigger a close event of an open snackbar that could be handled
+          // after openDialog, so the (re-)opening of the snackbar is delayed to avoid close being handled after open.
+          process.nextTick(() =>
+            dispatch({ type: 'openSnackbar', payload: 'noPortfolios' })
+          );
+        }
+      }
+    } catch (e) {
+      if (isMounted.current) {
+        dispatch({ type: 'setError', payload: e });
+        process.nextTick(() =>
+          dispatch({ type: 'openSnackbar', payload: 'error' })
+        );
+      }
+    } finally {
+      dispatch({ type: 'setLoading', payload: false });
+    }
+  };
+
+  return (
+    <div>
+      <ProgressButton
```

Comment:
Why is this a ProgressButton?

Reply:
The portfolios with stock quantities are loaded when the button is clicked (immediately before opening the dialog), not when the component is rendered. To indicate the loading progress, the `ProgressButton` component is used.

The most important reasons I decided against loading this information on first render:

* The button may never be clicked. The request would be unnecessary in that case.
* If loading fails, an error would appear "out of nowhere" or it would fail silently, both not good imo.
* The information could be outdated.

Reply:
I always thought there was going to be a loading bar in the dialog itself, but now that it's a shared component and the edit button doesn't need to init load, this probably makes the most sense.